pkg/provisioning/service.go

```diff
@@ -72,32 +76,47 @@ func (s *Service) Init(ctx context.Context) error {
 		Str("pipelines_path", s.pipelinesPath).
 		Msg("initializing the provisioning service")
 
-	var files []string
-	err := filepath.WalkDir(s.pipelinesPath, func(path string, fileInfo fs.DirEntry, err error) error {
-		if strings.HasSuffix(path, ".yml") || strings.HasSuffix(path, ".yaml") {
-			files = append(files, path)
-		}
-		return nil
-	})
+	files, err := s.getYamlFiles(s.pipelinesPath)
 	if err != nil {
-		return cerrors.Errorf("could not iterate through the pipelines folder %q: %w", s.pipelinesPath, err)
+		return cerrors.Errorf("failed to read pipelines folder %q: %w", s.pipelinesPath, err)
 	}
 
 	// contains all the errors occurred while provisioning configuration files.
 	var multierr error
-	// contains pipelineIDs of successfully provisioned pipelines.
-	var successPls []string
-	// contains pipelineIDs of all the pipelines in all the configuration files, either successfully provisioned or not.
-	var allPls []string
+
+	// parse pipeline config files
+	configs := make([]config.Pipeline, 0, len(files))
 	for _, file := range files {
-		provPipelines, all, err := s.provisionConfigFile(ctx, file, successPls)
+		cfg, err := s.parsePipelineConfigFile(ctx, file)
 		if err != nil {
-			multierr = multierror.Append(multierr, err)
+			return multierror.Append(multierr, err)
 		}
-		allPls = append(allPls, all...)
-		if len(provPipelines) != 0 {
-			successPls = append(successPls, provPipelines...)
+		configs = append(configs, cfg...)
+	}
+
+	// contains pipelineIDs of all the pipelines in all the configuration files, either successfully provisioned or not.
+	var allPls []string
+
+	// delete duplicate pipelines (if any)
+	for duplicateID, duplicateIndexes := range s.findDuplicateIDs(configs) {
+		multierr = cerrors.Errorf("%d pipelines with ID %q will be skipped: %w", len(duplicateIndexes), duplicateID, ErrDuplicatedPipelineID)
+		configs = s.deleteIndexes(configs, duplicateIndexes)
+
+		// duplicated IDs should still count towards all encountered pipeline IDs
+		allPls = append(allPls, duplicateID)
+	}
```

Comment:
I think we didn't discuss it explicitly and just said it's _good enough_. Given that the behavior is non-deterministic (i.e. you don't know which pipeline will be provisioned and which one won't be), I think it's better to change it and not provision anything in case of a conflict.

Pipelines in the configuration file are in a list, yes. Here we detect duplicates even across files.